StreamingASR/StreamingASR/app/src/main/java/org/pytorch/demo/streamingasr/MainActivity.java

```diff
@@ -0,0 +1,388 @@
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            requestPermissions(
+                    new String[]{android.Manifest.permission.RECORD_AUDIO}, REQUEST_RECORD_AUDIO);
+        }
+    }
+
+    private String assetFilePath(Context context, String assetName) {
+        File file = new File(context.getFilesDir(), assetName);
+        if (file.exists() && file.length() > 0) {
+            return file.getAbsolutePath();
+        }
+
+        try (InputStream is = context.getAssets().open(assetName)) {
+            try (OutputStream os = new FileOutputStream(file)) {
+                byte[] buffer = new byte[4 * 1024];
+                int read;
+                while ((read = is.read(buffer)) != -1) {
+                    os.write(buffer, 0, read);
+                }
+                os.flush();
+            }
+            return file.getAbsolutePath();
+        } catch (IOException e) {
+            Log.e(TAG, assetName + ": " + e.getLocalizedMessage());
+        }
+        return null;
+    }
+
+    private void showTranslationResult(String result) {
+        mTextView.setText(result);
+    }
+
+    public void run() {
+        android.os.Process.setThreadPriority(android.os.Process.THREAD_PRIORITY_AUDIO);
+
+        int bufferSize = AudioRecord.getMinBufferSize(SAMPLE_RATE, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT);
+        AudioRecord record = new AudioRecord(MediaRecorder.AudioSource.DEFAULT, SAMPLE_RATE, AudioFormat.CHANNEL_IN_MONO, AudioFormat.ENCODING_PCM_16BIT,
+                bufferSize);
+
+        if (record.getState() != AudioRecord.STATE_INITIALIZED) {
+            Log.e(TAG, "Audio Record can't initialize!");
+            return;
+        }
+        record.startRecording();
+
+        int chunkToRead = CHUNK_TO_READ;
+        int recordingOffset = 0;
+        short[] recordingBuffer = new short[CHUNK_TO_READ*CHUNK_SIZE];
+        while (mListening) {
+            long shortsRead = 0;
+            short[] audioBuffer = new short[bufferSize / 2];
+
+            while (shortsRead < chunkToRead * CHUNK_SIZE) {
+                // for every segment of 5 chunks of data, we perform transcription
+                // each successive segment’s first chunk is exactly the preceding segment’s last chunk
+                int numberOfShort = record.read(audioBuffer, 0, audioBuffer.length);
+                shortsRead += numberOfShort;
+                int x = (int) (numberOfShort - (shortsRead - chunkToRead * CHUNK_SIZE));
+                if (shortsRead > chunkToRead * CHUNK_SIZE)
+                    System.arraycopy(audioBuffer, 0, recordingBuffer, recordingOffset, (int) (numberOfShort - (shortsRead - chunkToRead*640)));
+                else
+                    System.arraycopy(audioBuffer, 0, recordingBuffer, recordingOffset, numberOfShort);
+
+                recordingOffset += numberOfShort;
+            }
+            double[] floatInputBuffer = new double[CHUNK_TO_READ * CHUNK_SIZE];
```

Comment:
Allocated in every cycle of `while`, can it be allocated only once and reused?

Reply:
yes it can.